tools/c7n_traildb/trailes.py

```diff
@@ -0,0 +1,277 @@
+                    'name': {'type': 'string'},
+                    'title': {'type': 'string'},
+                    'tags': {'type': 'object'},
+                    'bucket': {'type': 'string'},
+                    'regions': {'type': 'array', 'items': {'type': 'string'}}
+                }
+            }
+        }
+    }
+}
+
+
+class MyConnection(RequestsHttpConnection):
+    def __init__(self, *args, **kwargs):
+        proxies = kwargs.pop('proxies', {})
+        super(MyConnection, self).__init__(*args, **kwargs)
+        self.session.proxies = proxies
+
+
+def get_es_client(config):
+    host = [config['indexer'].get('host', 'localhost')]
+    es_kwargs = {}
+    user = config['indexer'].get('user', False)
+    password = config['indexer'].get('password', False)
+    if user and password:
+        es_kwargs['http_auth'] = (user, password)
+
+    es_kwargs['port'] = config['indexer'].get('port', 9200)
+
+    proxy = config['indexer'].get('proxy', False)
+    if proxy:
+        es_kwargs['connection_class'] = MyConnection
+        es_kwargs['proxies'] = {'https': proxy}
+
+    return Elasticsearch(host, **es_kwargs)
+
+
+def index_events(client, events):
+    results = helpers.streaming_bulk(client, events)
+    for status, r in results:
+        if not status:
+            log.debug("index err result %s", r)
+
+
+def dict_factory(cursor, row):
+    """Returns a sqlite row factory that returns a dictionary"""
+    d = {}
+    for idx, col in enumerate(cursor.description):
+        d[col[0]] = row[idx]
+    return d
+
+
+def fetch_events(cursor, query, config, account_name):
+    """Generator that returns the events"""
+    for event in cursor.execute(query):
+        event['account'] = account_name
+        event['_index'] = config['indexer']['idx_name']
+        event['_type'] = config['indexer'].get('idx_type', 'traildb')
+        yield event
+
+
+def get_traildb(bucket, key, session_factory, directory):
+    local_db_file = directory + "/traildb" + \
+        str(thread.get_ident())
+    local_bz2_file = local_db_file + '.bz2'
+
+    s3 = local_session(session_factory).resource('s3')
+    s3.Bucket(bucket).download_file(key['Key'], local_bz2_file)
+
+    # Decompress the traildb file
+    # (use the system bunzip2 command because it's faster than the python bzip2 library)
+    subprocess.call(['bunzip2', '-f', local_bz2_file])
+
+    return local_db_file
+
+
+def valid_date(key, config_date):
+    """ traildb bucket folders are not zero-padded so this validation
+        checks that the keys returned by the paginator are
+        *really after* the config date
+    """
+    key_date = "/".join(key.split("/")[4:7])
+    return parse_date(key_date) > parse_date(config_date)
+
+
+def index_account_trails(config, account, region, date, directory):
+    es_client = get_es_client(config)
+
+    s3 = local_session(
+        lambda : assumed_session(account['role'], 'TrailIndex')).client('s3')
+
+    bucket = account['bucket']
```

Comment:
there's enough duplication / copy and paste from trailts thinking  we should abstract the common to a module.